File: pythonDemos/body.py
import math
import pygame
from quad import Quad

class Body:
    __slots__ = ['_rx', '_ry', '_vx', '_vy', '_mass', '_color', '_fx', '_fy', 'radius','overRiddenRadius']
    G:float = 6.67430e-11
    color:tuple[int] = (0,0,0)
    elastic:float = 1.0
    EPSILON:float = 10**-4
    def __init__(self, rx:float,ry:float,vx:float, vy:float,mass:float,color:tuple[int]):
        self._mass = mass
        self._rx = rx
        self._ry = ry
        self._vx = vx
        self._vy = vy
        self._color = color
        self._fx = 0.0
        self._fy = 0.0
        self.radius = self._mass ** (1/3)
        self.overRiddenRadius:int = None
    
    @property
    def scaledRadius(self)->int:
        return min(max(1,int(self.radius/1000)),25) if self.overRiddenRadius == None else self.overRiddenRadius
    @property
    def copy(self)->'Body':
        '''return a copy of the body'''
        newColor = (*self._color,) #defensive copy
        return Body(self._rx, self._ry, self._vx, self._vy, self._mass, newColor)
    
    def update(self, dt:float)->None:
        #rk4
        k1 = self._fx / self._mass
        l1 = self._fy / self._mass
        k2 = (self._fx + 0.5 * k1) / self._mass
        l2 = (self._fy + 0.5 * l1) / self._mass
        k3 = (self._fx + 0.5 * k2) / self._mass
        l3 = (self._fy + 0.5 * l2) / self._mass
        k4 = (self._fx + k3) / self._mass
        l4 = (self._fy + l3) / self._mass
        self._vx += dt * (k1 + 2 * k2 + 2 * k3 + k4) / 6
        self._vy += dt * (l1 + 2 * l2 + 2 * l3 + l4) / 6
        self._rx += dt * self._vx
        self._ry += dt * self._vy

    def distanceTo(self, b:'Body')->float:
        dx = b._rx - self._rx
        dy = b._ry - self._ry
        return math.sqrt(dx*dx + dy*dy)
    
    def resetForce(self)->None:
        self._fx = 0.0
        self._fy = 0.0

    def addForce(self, b:'Body')->None:
        # EPS = 3E4
        dx = b._rx - self._rx
        dy = b._ry - self._ry
        dist = max(math.sqrt(dx*dx + dy*dy),self.scaledRadius+b.scaledRadius)
        F = (Body.G * self._mass * b._mass) / (dist*dist)
        self._fx += F * dx / dist
        self._fy += F * dy / dist

    def inQuad(self,q:'Quad')->bool:
        return q.contains(self._rx, self._ry)
    
    def plus(self, b:'Body')->'Body':
        mass = self._mass + b._mass
        rx = (self._rx * self._mass + b._rx * b._mass) / mass
        ry = (self._ry * self._mass + b._ry * b._mass) / mass
        vx = (self._vx * self._mass + b._vx * b._mass) / mass
        vy = (self._vy * self._mass + b._vy * b._mass) / mass
        return Body(rx, ry, vx, vy, mass, self._color)

    def collide(self,other:'Body')->None:
        self.__collide(self,other)
    
    @staticmethod
    def __collide(b1:'Body',b2:'Body')->None:
    # def collide(self, b1:Body, b2:Body)->None:
        """
        Handles an elastic or inelastic collision between two bodies.
        The `elastic` coefficient (1.0 = perfectly elastic, 0.0 = perfectly inelastic) determines energy loss.
        """
        # print("colliding")
        # print(b1)
        # print(b2)
        # Extract properties
        m1, m2 = b1._mass, b2._mass
        r1, r2 = b1.scaledRadius, b2.scaledRadius
        x1, y1, x2, y2 = b1._rx, b1._ry, b2._rx, b2._ry
        vx1, vy1, vx2, vy2 = b1._vx, b1._vy, b2._vx, b2._vy
        e = Body.elastic  # Elasticity coefficient

        # Compute relative position and distance
        dx, dy = x2 - x1, y2 - y1
        dist = max((dx**2 + dy**2) ** 0.5, 1e-10)  # Avoid division by zero

        # Normal and tangent vectors
        nx, ny = dx / dist, dy / dist
        tx, ty = -ny, nx

        # Decompose velocities
        dpTan1 = vx1 * tx + vy1 * ty
        dpTan2 = vx2 * tx + vy2 * ty
        dpNorm1 = vx1 * nx + vy1 * ny
        dpNorm2 = vx2 * nx + vy2 * ny

        # Compute new normal velocities using elastic coefficient
        m1f = ((m1 - e * m2) * dpNorm1 + (1 + e) * m2 * dpNorm2) / (m1 + m2)
        m2f = ((m2 - e * m1) * dpNorm2 + (1 + e) * m1 * dpNorm1) / (m1 + m2)

        # Convert back to x, y components
        b1._vx, b1._vy = tx * dpTan1 + nx * m1f, ty * dpTan1 + ny * m1f
        b2._vx, b2._vy = tx * dpTan2 + nx * m2f, ty * dpTan2 + ny * m2f

        # Resolve overlap by shifting bodies apart in a mass-weighted manner
        overlap = r1 + r2 - dist
        if overlap > 0:
            # correction1 = (m2/ (m1 + m2)) * overlap
            # correction2 = (m1/ (m1 + m2)) * overlap
            # b1._rx -= correction1 * nx
            # b1._ry -= correction1 * ny
            # b2._rx += correction2 * nx
            # b2._ry += correction2 * ny
            # b1._rx -= overlap * nx
            # b1._ry -= overlap * ny
            # b2._rx += overlap * nx
            # b2._ry += overlap * ny



            largeBody = b1 if b1._mass > b2._mass else b2
            smallBody = b1 if largeBody == b2 else b2
            # smallbody get's shifted by the overlap, big one doesn't.
            dx = largeBody._rx - smallBody._rx
            dy = largeBody._ry - smallBody._ry
            dist = math.sqrt(dx*dx + dy*dy)+Body.EPSILON
            nx = dx / dist
            ny = dy / dist

            overlap = largeBody.scaledRadius + smallBody.scaledRadius - dist
            smallBody._rx -= overlap * nx
            smallBody._ry -= overlap * ny

            # smallBody._rx += overlap * nx
            # smallBody._ry += overlap * ny      

        
    def draw(self, screen:pygame.display)->None:
        position = (round(self._rx), round(self._ry))
        pygame.draw.circle(screen, self._color, (*position,), self.scaledRadius)
    
    def __str__(self)->str:
        return f"Body(rx={self._rx}, ry={self._ry}, vx={self._vx}, vy={self._vy}, mass={self._mass}, color={self._color}, hash={self.__hash__()})"
    
    def __repr__(self)->str:
        return self.__str__()
    
    def __equals__(self, b:'Body')->bool:
        return self.__hash__() == b.__hash__()
    

File: src/Body.java
import java.awt.Color;
import java.awt.Graphics;

public class Body{
    public double G = 6.6743e-11;
    public double elastic = 1.0;
    public static double EPSILON = 10e-4;

    private double rx, ry;
    private double vx, vy;
    private double mass;
    private Color color;
    public Color SwitchColor = Color.YELLOW;
    private double radius;
    private double fx, fy;
    public Integer overRiddenRadius;
    public boolean collided = false;
    public boolean changeColorOnCollision = false;

    public Body(double rx, double ry, double vx, double vy, double mass, Color color){
        this.rx = rx;
        this.ry = ry;
        this.vx = vx;
        this.vy = vy;
        this.mass = mass;
        this.color = color;
        this.fx = 0.0;
        this.fy = 0.0;
        this.radius = Math.pow(mass, 1.0/3.0);
        this.overRiddenRadius = null;
    }

    public int scaledRadius(){
        if(this.overRiddenRadius == null){
            return (int) Math.min(Math.max(1.0, this.radius/1000.0),25.0);
        }else{
            return this.overRiddenRadius.intValue();
        }
    }

    public Body copy(){
        return new Body(rx, ry, vx, vy, mass, new Color(color.getRGB()));
    }

    public void update(double dt){
        /*
         * commented out runhe kutta integration because it makes small particles
         * accelerate differenelty to larger particles. 
         * I don't really care too much about the accuracy of the integration in the gravity sims to sacrifice having realistic physics for small particle sims.
         * so I'm just using good ole newtonian mechanics equations. And assuming constant acceleration during the frame.
         * 
         */
        // double k1 = this.fx / this.mass;
        // double l1 = this.fy / this.mass;
        // double k2 = (this.fx + 0.5 * k1) / this.mass;
        // double l2 = (this.fy + 0.5 * l1) / this.mass;
        // double k3 = (this.fx + 0.5 * k2) / this.mass;
        // double l3 = (this.fy + 0.5 * l2) / this.mass;
        // double k4 = (this.fx + k3) / this.mass;
        // double l4 = (this.fy + l3) / this.mass;
        // this.vx += dt * (k1 + 2 * k2 + 2 * k3 + k4) / 6;
        // this.vy += dt * (l1 + 2 * l2 + 2 * l3 + l4) / 6;
        // this.rx += dt * this.vx;
        // this.ry += dt * this.vy;

        // assuming constant acceleration
        // s = ut + 1/2at^2

        // travelled during frame
        this.rx += dt * this.vx + 0.5 * dt * dt * this.fx / this.mass;
        this.ry += dt * this.vy + 0.5 * dt * dt * this.fy / this.mass;

        // accelerated during frame
        this.vx += dt * this.fx / this.mass;
        this.vy += dt * this.fy / this.mass;

    }

    public double distanceTo(Body b){
        double dx = this.rx - b.rx;
        double dy = this.ry - b.ry;
        return Math.sqrt(dx*dx + dy*dy);
    }

    public void resetForce(){
        this.fx = 0.0;
        this.fy = 0.0;
    }

    public void addForce(Body b){
        double dx = b.rx - this.rx;
        double dy = b.ry - this.ry;
        double dist = Math.max(this.distanceTo(b), this.scaledRadius() + b.scaledRadius());
        double F = (G * this.mass * b.mass) / (dist * dist);
        this.fx += F * dx / dist;
        this.fy += F * dy / dist;
    }

    public void addForce(double fx, double fy){
        this.fx += fx;
        this.fy += fy;
    }

    public boolean inQuad(Quad q){
        return q.contains(this.rx, this.ry);
    }

    public Body plus(Body b){
        double m = this.mass + b.mass;
        double _rx = (this.rx * this.mass + b.rx * b.mass) / m;
        double _ry = (this.ry * this.mass + b.ry * b.mass) / m;
        double _vx = (this.vx * this.mass + b.vx * b.mass) / m;
        double _vy = (this.vy * this.mass + b.vy * b.mass) / m;
        return new Body(_rx, _ry, _vx, _vy, m, new Color(this.color.getRGB()));
    }

    public void collide(Body other){
        privateCollide(this,other);
        this.collided = true;
    }

    public static void privateCollide(Body b1, Body b2){
        double m1 = b1.mass;
        double m2 = b2.mass;
        double r1 = b1.scaledRadius();
        double r2 = b2.scaledRadius();
        
        double x1 = b1.rx;
        double y1 = b1.ry;
        double x2 = b2.rx;
        double y2 = b2.ry;

        double vx1 = b1.vx;
        double vy1 = b1.vy;
        double vx2 = b2.vx;
        double vy2 = b2.vy;

        double e = (b1.elastic+b2.elastic)/2;

        //relative position
        double dx = x2 - x1;
        double dy = y2 - y1;
        double dist = Math.max(Math.sqrt(dx*dx + dy*dy), Body.EPSILON);

        //Normal and Tangent vectors
        double nx = dx/dist;
        double ny = dy/dist;
        double tx = -ny;
        double ty = nx;

        //decompose velocities
        double dpTan1 = vx1 * tx + vy1 * ty;
        double dpTan2 = vx2 * tx + vy2 * ty;
        double dpNorm1 = vx1 * nx + vy1 * ny;
        double dpNorm2 = vx2 * nx + vy2 * ny;

        //new normal velocities using elastic collision equation
        double m1f = ((m1 - e * m2) * dpNorm1 + (1 + e) * m2 * dpNorm2) / (m1 + m2);
        double m2f = ((m2 - e * m1) * dpNorm2 + (1 + e) * m1 * dpNorm1) / (m1 + m2);

        //back to original vector basis
        b1.vx = tx * dpTan1 + nx * m1f;
        b1.vy = ty * dpTan1 + ny * m1f;
        b2.vx = tx * dpTan2 + nx * m2f;
        b2.vy = ty * dpTan2 + ny * m2f;

        //move bodies so they don't overlap
        double overlap = r1 + r2 - dist;
        if(overlap > 0){
            //small body gets shifted. Not the large body. Otherwise you have dust pushing suns around.
            Body largeBody = m1 > m2 ? b1 : b2;
            Body smallBody = m1 > m2 ? b2 : b1;

            dx = largeBody.rx - smallBody.rx;
            dy = largeBody.ry - smallBody.ry;
            dist = Math.max(Math.sqrt(dx*dx + dy*dy), Body.EPSILON);
            nx = dx/dist;
            ny = dy/dist;

            overlap = largeBody.scaledRadius() + smallBody.scaledRadius() - dist;
            smallBody.rx -= nx * overlap / 2;
            smallBody.ry -= ny * overlap / 2;

        }
    }

    public void draw(Graphics g){
        int posX = (int) Math.round(this.rx);
        int posY = (int) Math.round(this.ry);
        Color _color = (this.changeColorOnCollision && this.collided) ? this.SwitchColor : this.color;

        g.setColor(_color);
        g.fillOval(posX - this.scaledRadius(), posY - this.scaledRadius(), 2 * this.scaledRadius(), 2 * this.scaledRadius());
        this.collided = false;

    }
    //getters
    public double getX(){
        return this.rx;
    }
    public double getY(){
        return this.ry;
    }
    public double getVx(){
        return this.vx;
    }
    public double getVy(){
        return this.vy;
    }
    public double getMass(){
        return this.mass;
    }
    //setters
    public void setPosition(double x, double y){
        this.rx = x;
        this.ry = y;
    }
    public void setVelocity(double vx, double vy){
        this.vx = vx;
        this.vy = vy;
    }


}